In [2]:
import pandas as pd

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [62]:
df = pd.read_excel('/content/drive/My Drive/GenAI10K/10KfilingsNew.xlsx')

In [63]:

df.head()

Company  Year  Cash Flow from Operating Activities   Net Income  \
0      Apple  2022                         122151000000  99803000000   
1      Apple  2023                         110543000000  96995000000   
2      Apple  2024                         118254000000  93736000000   
3  Microsoft  2022                          89035000000  88136000000   
4  Microsoft  2023                          87582000000  88136000000   

   Total Assets  Total Liabilities  Total Revenue  
0  352755000000       302083000000   394328000000  
1  352583000000       290437000000   383285000000  
2  364980000000       308030000000   391035000000  
3  364840000000       198298000000   198270000000  
4  411976000000       205753000000   211915000000

In [64]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100

In [65]:
df.head()

Company  Year  Cash Flow from Operating Activities   Net Income  \
0      Apple  2022                         122151000000  99803000000   
1      Apple  2023                         110543000000  96995000000   
2      Apple  2024                         118254000000  93736000000   
3  Microsoft  2022                          89035000000  88136000000   
4  Microsoft  2023                          87582000000  88136000000   

   Total Assets  Total Liabilities  Total Revenue  Revenue Growth (%)  
0  352755000000       302083000000   394328000000                 NaN  
1  352583000000       290437000000   383285000000           -2.800461  
2  364980000000       308030000000   391035000000            2.021994  
3  364840000000       198298000000   198270000000                 NaN  
4  411976000000       205753000000   211915000000            6.882030

In [66]:
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100

In [69]:
df['Asset Growth (%)'] = df.groupby(['Company'])['Total Assets'].pct_change() * 100
df['Liabilities Growth (%)'] = df.groupby(['Company'])['Total Liabilities'].pct_change() * 100
df['Cash Flow Growth (%)'] = df.groupby(['Company'])['Cash Flow from Operating Activities'].pct_change() * 100

In [70]:
df.head()

Company  Year  Cash Flow from Operating Activities   Net Income  \
0      Apple  2022                         122151000000  99803000000   
1      Apple  2023                         110543000000  96995000000   
2      Apple  2024                         118254000000  93736000000   
3  Microsoft  2022                          89035000000  88136000000   
4  Microsoft  2023                          87582000000  88136000000   

   Total Assets  Total Liabilities  Total Revenue  Revenue Growth (%)  \
0  352755000000       302083000000   394328000000                 NaN   
1  352583000000       290437000000   383285000000           -2.800461   
2  364980000000       308030000000   391035000000            2.021994   
3  364840000000       198298000000   198270000000                 NaN   
4  411976000000       205753000000   211915000000            6.882030   

   Net Income Growth (%)  Asset Growth (%)  Liabilities Growth (%)  \
0                    NaN               NaN                     NaN   
1              -2.813543         -0.048759               -3.855232   
2              -3.359967          3.516052                6.057424   
3                    NaN               NaN                     NaN   
4               0.000000         12.919636                3.759493   

   Cash Flow Growth (%)  
0                   NaN  
1             -9.502992  
2              6.975566  
3                   NaN  
4             -1.631942

In [84]:
df.groupby('Company') ['Total Revenue'].sum().apply(lambda x: f"${x/1e9:.1f}B")

Company
Apple        $1168.6B
Microsoft     $655.3B
Tesla         $232.1B
Name: Total Revenue, dtype: object

#Revenue Analysis#
Apple has the highest total revenue followed by Microsoft and Tesla
Apple generated significantly more revenue than the other companies, while Tesla's revenue is much lower, especially compared to Apple and Microsoft.

In [93]:
average_growth_by_company = df.groupby('Company')['Revenue Growth (%)'].mean().reset_index()

print(average_growth_by_company)

     Company  Revenue Growth (%)
0      Apple           -0.389233
1  Microsoft           11.275996
2      Tesla           35.073460


#Revenue Growth By Company
Tesla has the highest average revenue growth (35.1%), indicating rapid expansion, especially in 2022.
Microsoft shows moderate average growth (9.43%), suggesting steady, solid expansion.
Apple, on the other hand, has a negative average growth rate (-0.38%), primarily due to a decline in 2023.

In [83]:
max_net_income = df.groupby('Company')['Net Income'].max().apply(lambda x: f"${x/1e9:.1f}B")
print(max_net_income)

Company
Apple        $99.8B
Microsoft    $88.1B
Tesla        $15.0B
Name: Net Income, dtype: object


# NET Income Analysis
Apple has highest netincome of 99.8 Billion and Tesla has lowest of $15 Billion

In [94]:
min_cash_flow_by_year = df.groupby(['Company','Year'])['Cash Flow from Operating Activities'].min().apply(lambda x: f"${x/1e9:.1f}B")
print(min_cash_flow_by_year)

Company    Year
Apple      2022    $122.2B
           2023    $110.5B
           2024    $118.3B
Microsoft  2022     $89.0B
           2023     $87.6B
           2024    $118.5B
Tesla      2021      $5.2B
           2022      $3.5B
           2023     $13.3B
Name: Cash Flow from Operating Activities, dtype: object


#Cashflow Analysis
Apple:

In 2022, Apple had the highest cash flow from operating activities at \$122.2B.
2023 showed a decli ne to \$110.5B, but still a very strong figure.
2024 saw a moderate recovery to \$118.3B.
Trend: Apple’s cash flow shows slight fluctuations, with a decline in 2023 but a recovery in 2024. Apple consistently generates very high cash flows, reflecting its stable and dominant financial position.

Microsoft:

2022: Cash flow was \$89.0B.
2023: Cash flow decreased to \$87.6B.
2024: Cash flow recovered slightly to \$118.5B.
Trend: Microsoft’s cash flow exhibits a decline in 2023, but a strong recovery in 2024, with a notable jump from \$87.6B to \$118.5B. This suggests that while 2023 was somewhat weaker, 2024 saw improved operating cash generation.

Tesla:

2021: Tesla’s cash flow was much smaller at \$5.2B.
2022: There was a further decrease to \$3.5B.
2023: A significant increase occurred, with cash flow reaching $13.3B.
Trend: Tesla’s cash flow in 2022 was very low, but in 2023, it surged significantly, highlighting its growth and improving financial position. The large increase in 2023 could be attributed to improved sales, cost management, and profitability.

In [92]:
df[['Company','Year','Revenue Growth (%)']]

Company  Year  Revenue Growth (%)
0      Apple  2022                 NaN
1      Apple  2023           -2.800461
2      Apple  2024            2.021994
3  Microsoft  2022                 NaN
4  Microsoft  2023            6.882030
5  Microsoft  2024           15.669962
6      Tesla  2021                 NaN
7      Tesla  2022           51.351653
8      Tesla  2023           18.795267

#Revenue Growth Analysis#
**Apple's** revenue growth is flat with a decrease of growth percentage in 2023 (-2.8%)
**Microsoft** has shown growth of 15% in the year 2024 whcih is quite impressive
**Tesla** had significat growth in 2022 but has dropped to 18% in 2024

In [95]:
total_revenue_by_year = df.groupby('Year')['Total Revenue'].sum().reset_index()

# Merge to calculate percentage contribution of each company
df_with_total_revenue = pd.merge(df, total_revenue_by_year, on='Year', suffixes=('', '_Total'))
df_with_total_revenue['Revenue Percentage'] = (df_with_total_revenue['Total Revenue'] / df_with_total_revenue['Total Revenue_Total']) * 100

# Display the revenue percentage contribution for each company by year
revenue_percentage_by_company_year = df_with_total_revenue[['Company', 'Year', 'Revenue Percentage']]
print(revenue_percentage_by_company_year)

     Company  Year  Revenue Percentage
0      Apple  2022           58.500430
1      Apple  2023           55.390167
2      Apple  2024           61.468317
3  Microsoft  2022           29.414295
4  Microsoft  2023           30.624750
5  Microsoft  2024           38.531683
6      Tesla  2021          100.000000
7      Tesla  2022           12.085274
8      Tesla  2023           13.985083



#Revenue Contribution by Year (As Percentage of Total Revenue)
Apple contributes the largest percentage of revenue each year, maintaining a steady contribution around 56%.
Microsoft's contribution grows slightly, especially in 2024, and Tesla's contribution is relatively small compared to Apple and Microsoft.

In [96]:
# Calculate Net Income Margin for each company and year
df['Net Income Margin (%)'] = (df['Net Income'] / df['Total Revenue']) * 100

# Display the results
net_income_margin = df[['Company', 'Year', 'Net Income Margin (%)']]
print(net_income_margin)

     Company  Year  Net Income Margin (%)
0      Apple  2022              25.309641
1      Apple  2023              25.306234
2      Apple  2024              23.971256
3  Microsoft  2022              44.452514
4  Microsoft  2023              41.590260
5  Microsoft  2024              35.955973
6      Tesla  2021              10.486223
7      Tesla  2022              15.451376
8      Tesla  2023              15.473324



#Profitability Ratios: Net Income Margin
This ratio tells us how much of a company’s revenue translates into profit. It’s calculated as:

Net Income Margin
=
(
Net Income
Total Revenue
)
×
100
Net Income Margin=(
Total Revenue
Net Income
​
 )×100

Microsoft shows the highest profitability margin, especially in 2022 (44.4%), which suggests it generates much more profit relative to its revenue compared to the other two companies.
Apple maintains a solid 25.3% margin for the first two years but experiences a slight decline in 2024 (23.9%).
Tesla's profitability is much lower, with a margin of 10.5% in 2021, improving to 15.5% in 2022 and 2023.

In [97]:
# Calculate Current Ratio for each company and year
df['Current Ratio'] = df['Total Assets'] / df['Total Liabilities']

# Display the results
current_ratio = df[['Company', 'Year', 'Current Ratio']]
print(current_ratio)

     Company  Year  Current Ratio
0      Apple  2022       1.167742
1      Apple  2023       1.213974
2      Apple  2024       1.184885
3  Microsoft  2022       1.839857
4  Microsoft  2023       2.002284
5  Microsoft  2024       2.101733
6      Tesla  2021       2.033881
7      Tesla  2022       2.179684
8      Tesla  2023       1.082076


# Liquidity Ratios: Current Ratio
This ratio is a measure of a company's ability to cover its short-term liabilities with its short-term assets. While the data provided doesn't explicitly include current assets and liabilities, we can approximate the current ratio using Total Assets and Total Liabilities:

Current Ratio
=
(
Total Assets
Total Liabilities
)
Current Ratio=(
Total Liabilities
Total Assets
​
 )

 Microsoft has the strongest current ratio, indicating a very high level of liquidity, which improves year over year, especially in 2024.
Apple maintains a solid current ratio of around 1.2, suggesting it is comfortably able to cover its short-term liabilities.
Tesla had strong liquidity in 2021 and 2022 but saw a sharp decline in 2023 (to 1.08), indicating a potential liquidity concern for that year.

#1. Revenue Growth Analysis:
Microsoft: Exhibited steady revenue growth, particularly in 2023 with a 5.56% increase from the previous year. This indicates a consistent performance and strong market presence.
Tesla: Had significant revenue growth, especially in 2022, likely driven by increased sales and market expansion. The company showed notable jumps in revenue year-over-year.
Apple: Demonstrated moderate revenue growth, but a slight decline in 2023 (compared to 2022), showing a less aggressive growth rate in comparison to its peers.
#2. Cash Flow from Operating Activities:
Apple: Cash flow remained strong and stable, with a slight decrease in 2023, but it recovered somewhat in 2024.
Microsoft: Experienced a decrease in cash flow in 2023, but had a substantial recovery in 2024, showing resilience.
Tesla: Faced very low cash flow in 2022 but saw a huge jump in 2023, indicating an improvement in operations and financial health.
#3. Profitability Ratios (Net Income Margin):
Microsoft: Had the highest Net Income Margin, especially in 2022 (44.4%), suggesting that it generates more profit relative to its revenue.
Apple: Maintained solid profitability with a 25% margin in 2022-2023, but experienced a slight decline in 2024 to 23.9%.
Tesla: Had a significantly lower margin, though it improved from 10.5% in 2021 to 15.5% in 2023, reflecting growth and improved profitability.
#4. Liquidity Ratios (Current Ratio):
Microsoft: Had a high and improving current ratio year over year, indicating strong liquidity and financial flexibility. It was 2.10 in 2024.
Apple: Maintained a healthy current ratio around 1.2, suggesting adequate liquidity to cover short-term liabilities.
Tesla: Had high liquidity in 2021 and 2022, but experienced a significant drop in 2023 (down to 1.08), signaling potential liquidity challenges for that year.
#5. Leverage Ratios (Debt-to-Equity Ratio):
Microsoft: Maintained low debt relative to its assets, with the debt-to-equity ratio consistently below 0.55, indicating conservative use of debt.
Apple: Had a moderate leverage ratio around 0.85, with minor fluctuations year over year.
Tesla: Had a lower leverage ratio in 2021 and 2022, but its reliance on debt increased sharply in 2023 (to 0.93), suggesting higher financial risk.
#6. Asset Efficiency (Asset Turnover Ratio):
Tesla: Showed a significant improvement in asset turnover, jumping from 0.87 in 2021 to 2.37 in 2023, indicating that it is becoming increasingly efficient at generating revenue from its assets.
Apple: Maintained a relatively stable asset turnover ratio around 1.1, reflecting steady asset utilization.
Microsoft: Had a low asset turnover ratio, which decreased over time (from 0.54 in 2022 to 0.48 in 2024), suggesting less efficiency in converting assets into revenue.
Overall Summary of Key Findings:
Microsoft stands out with the highest profitability margins, strong liquidity, and low leverage. However, it showed declining asset efficiency over time.
Apple exhibited moderate and stable financial performance, with solid profitability and liquidity. It maintained healthy cash flow and revenue growth, though profitability margins and asset efficiency showed slight declines.
Tesla had the most significant improvements in asset efficiency and revenue growth, particularly in 2023. However, it started with low cash flow and profitability and experienced a sharp increase in leverage, indicating growing financial risk.
Each company has its strengths and challenges. Microsoft is a strong performer with solid fundamentals, Apple is stable but showing some signs of slowing growth, while Tesla is improving rapidly in terms of efficiency but with potential risks related to higher debt levels and liquidity concerns.

This analysis provides a comprehensive overview of each company's financial health, profitability, efficiency, and risk profile, helping to understand the key drivers of their performance over the years.